bip-taro.mediawiki

```diff
@@ -219,21 +225,134 @@ Taro-specific commitments, then the same or different leaf version can be used
+*** Or, is a tap leaf, as verifiable via the unique BIP-340 tagged hash tag.
+**** In this case, the leaf MUST be verified against the raw message digest to detect and reject duplicate Taro commitment inclusion in the tapscript tree (see the <code>taro_marker</code>).
+
+Due to the lexicographical sorting of tap branches/leaves before hashing, we
+lose the ordering information within the merkle tree. As a result, we're
+instead forced to require reveal of the sibling pre-image.
+
+The following algorithm can be used to asset the uniqueness of the Taro
+commitment in a tapscript tree:
+<source lang="python">
+taro_commitment_is_valid(commitment_output: TxOut, control_block: ControlBlock, 
+    sibling_preimage: TaroTapReveal, taro_root: []byte) -> bool
+    
+    taro_root_hash = tagged_hash("TapLeaf", taro_root)
+    internal_key = parse_key(control_block.internal_key_bytes)
+
+    match len(control_block.size):
+       # Just the internal public key, so we can just verify the commitment below.
+       case 32:
+           expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || taro_root_hash)
+
+           return expected_output_key == witness_program(commitment_output.pk_script)
+
+       # More than one element in the tree, so need to verify each sub-case.
+       case 65:
+           
+           match sibling_preimage:
+               # The pre-image is a leaf, so verify the hashes match up.
+               case LeafReveal(leaf_bytes):
+                   leaf_hash = tagged_hash("TapLeaf", leaf_bytes)
+
+                   if leaf_hash != control_block.sibling_hash:
+                       return false
+
+               # The pre-image is itself a branch, so we need the two siblings
+               # used to construct the branch.
+               case BranchReveal(sibling_1, sibling_2):
+                   expected_branch_hash = tagged_hash("TapBranch", sort(sibling_1, sibling_2))
+
+                   if expected_branch_hash != construct.sibling_hash:
+                       return false
+
+               # We know the tree is well formed now, so we'll verify the root
+               # commitment.
+               root_hash = tagged_hash("TapBranch", sort(taro_root_hash, control_block.sibling_hash))
+
+               expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || root_hash)
+
+               return expected_output_key == witness_program(commitment_output.pk_script)
+
+       default:
+           return false
+
+</source>
+
+In addition to verifying that a commitment is unique within the constraints of
+a single taproot tree, we also need to verify that there is no similar
+commitment across the remaining outputs in the genesis transaction. We can do
+this by asserting that:
```

Comment:
IIUC this requires having the preimages for every (single tapscript element or top-level tapscript branch) to cover each output in the TX. This seems worth mentioning explicitly.

This also seems like it would interfere with TXs built interactively with multiple parties (for each output added to a PSBT, we would also need to propagate tapscript preimages). Not sure if we can avoid that though.

Reply:
> IIUC this requires having the preimages for every (single tapscript element or top-level tapscript branch) to cover each output in the TX

Correct. In the typical case, I expect genesis transactions to just be 1-in-1 transactions, as there's no need to make a change output. If there is a change output, then we only require that opening. 

Added a sentence below this mentioning that. This is intended to be covered in the code fragment below where we examine the commitment opening.